llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -761,5 +761,147 @@ getLargeGRFSplitter(ModuleDesc &&MD, bool EmitOnlyKernelsAsEntryPoints) {
     return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
 }
 
+namespace {
+// Data structure, which represent a combination of all possible optional
+// features used in a function.
+//
+// It has extra methods to be useable as a key in llvm::DenseMap.
+struct UsedOptionalFeatures {
+  SmallVector<int, 4> Aspects;
+  // TODO: extend this further with reqd-sub-group-size, reqd-work-group-size,
+  // double-grf and other properties
+
+  UsedOptionalFeatures() = default;
+
+  UsedOptionalFeatures(const Function *F) {
+    if (const MDNode *MDN = F->getMetadata("sycl_used_aspects")) {
+      auto ExtractIntegerFromMDNodeOperand = [=](const MDNode *N,
+                                                 unsigned OpNo) -> auto {
+        Constant *C =
+            cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
+        return C->getUniqueInteger().getSExtValue();
+      };
+
+      // !sycl_used_aspects is supposed to contain unique values, no duplicates
+      // are expected here
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+        Aspects.push_back(ExtractIntegerFromMDNodeOperand(MDN, I));
+      }
```

Comment:
`llvm::transform` from `STLExtras.h` is even better because it accepts a range. ;)

Depends on having `MDN->operands()` or a similar accessor, though.

Reply:
Switched to `llvm::transform` in b39798bc075020de5c2621e0254a8dac46466f6d